docs/internal/rfc/002rfc-transition-explorer.md

```diff
@@ -0,0 +1,323 @@
+  `TransitionExplorer`, described below.
+- Introduce a new module, `TransitionExplorer` that enables the interactive
+  exploration of the transition system.
+
+*NOTE*: The high-level sketch above assumes the new code organization proposed
+in [ADR 7][].
+
+[ADR 7]: https://github.com/informalsystems/apalache/tree/unstable/docs/src/adr/007adr-restructuring.md
+
+#### API
+
+The following is a rough sketch of the proposed API for the transition explorer.
+It attempts to give a highly abstracted interface, but in terms of existing data
+structures. Naturally, refinements and alterations are to be expected during
+implementation.
+
+```scala
+
+/** A State is a map from variables to values  */
+type State = Map[TlaEx, TlaEx]
+
+/** An execution is an alternating sequence of states and actions,
+ *  terminating with a state 
+ */
+type Execution = List[Either[State, TlaEx]]
+
+trait TransitionExplorer {
+
+  /** Reset the state of the explorer
+   *
+   * Returns the explorer to the same state as if the currently loaded model where
+   * freshly loaded. Used to restart exploration from a clean slate.
+   *
+   * [TRANS-EX.1::LOAD.1]
+   */
+  def reset(): Unit
+
+  /** Load a model for exploration
+   *
+   * If a model is already loaded, it will be replaced and the state of the exploration
+   * [[reset]].
+   *
+   * [TRANS-EX.1::QCHECK.1]
+   * [TRANS-EX.1::LOAD.1]
+   *
+   * @param spec the root TLA+ module 
+   * @param aux auxiliary modules that may be required by the root module
+   * @return `Left(LoadErr)` if parsing or loading the model from `spec` goes
+   *          wrong, or `Right(())` if the model is loaded successfully.
+   */
+  def loadModel(spec: String, aux: List[String] = List()): Either[LoadErr, Unit]
+
+  /**  The root module currently loaded in memory  */
+  def loadedModel: Option[TlaModule]
+
+  /** The initial states of the model
+   *
+   * Since the number of computable initial states can be infinite, an upper
+   * limit must be set.
+   *
+   * @params max the maximum number of initial states to return (default to 100)
+   * @params start the nth state to begin fetching from (defaults to 0)
+   * @return `Some(exprs)` where `exprs` are `n` computed initial states, with
+   *         `n` <= `max`, or `None` if no model is loaded.
+   */
+  def initialStates(max: Int = 100, start: Int = 0): Option[TlaEx]
+
+  /**  The initial state predicates given in the spec  */
+  def initialStatePredicates: Option[List[TlaEx]]
+
+  /** The action predicates defining "next" transitions given in the spec
+   *
+   */
+  def actions: Option[List[TlaEx]]
+
+  /** The invariants defined by the spec */
+  def invariants: Option[CheckerInputVC]
+
+  /** Setting constants will also reset the explorer, if an exploration is 
+   * underway
+   */
+  def initializeConstants(Map[TlaExp, TlaExp]): Either[LoadErr, Unit]
+
+  /** Gives a map of constants to their current values */
+  def constants: Option[Map[TlaEx, Option[TlaEx]]]
```

Comment:
My logic was that this would let the method serve both as a way of listing the current values of the constants as a way of listing which constants there are (in case none have been set). I see what you mean about the implication that some be left absent tho.

Do you think I should add two different methods like `constants: Option[List[TlaEx]]` and `constantValues: Option[Map[TlaEx, TlaEx]]`?

Reply:
Not sure you need a method for that, `constants` would just be `constantValues map { _.keys }`

Reply:
Might be a slight misunderstanding. Of course if we are in our code base we don’t need a method for it. But the api here will be exposed via gRPC, and the clients will not have access to constantValues if we don’t explicitly then a method.

How this RFC is meant to be read is that the input types of the method signatures given here specify the requests the server will handle and the output types specify the replies the server will send in response.

I’ll think about what to add to make that clearer.

but the question here remains: do we want one endpoint for fetching the the constants and their values, or two: one to fetch just the constants (in case they aren’t set) and another to fetch their values if they have been set.

Reply:
I see what you mean. In my mind, the server should never be in a partially configured state, where it has received the spec, but the constants remain uninitialized. It should be possible to swap constant values, sure, but I would still require some constants be a part of the initialization. In that sense, there is only one option, fetching both the constants and the values.

Reply:
In that case, we would have to change the `loadSpec` endpoint to receive the constants values as well.

Reply:
Yes, but I think it's ultimately for the better.

Reply:
Cool. I've made that change, and I've also added a short, concrete example sketching out how the actual implementation will look using gRPC.